simulariumio/data_objects/binary_data.py

```diff
@@ -0,0 +1,131 @@
+from typing import Dict, List
+import json
+import numpy as np
+
+from .frame_data import DataIndices, FrameData, FrameMetadata
+from .input_file_data import InputFileData
+from .trajectory_data import TrajectoryData
+from .simularium_file_data import SimulariumFileData
+from ..constants import BINARY_BLOCK_TYPE, BINARY_SETTINGS
+from ..readers import BinaryBlockInfo, SimulariumBinaryReader
+
+
+class BinaryData(SimulariumFileData):
+    def __init__(self, file_name: str, file_contents: bytes):
+        """
+        This object holds binary encoded simulation trajectory file's
+        data while staying close to the original file format
+
+        Parameters
+        ----------
+        file_name : str
+            Name of the file
+        file_contents : bytes
+            A byte array containing the data of an open .simularium file
+        """
+        self.file_name = file_name
+        self.file_contents = InputFileData(file_contents=file_contents)
+        self.file_data = SimulariumBinaryReader._binary_data_from_source(
+            self.file_contents
+        )
+        self.frame_metadata: List[FrameMetadata] = []
+        self.block_indices: Dict[int, DataIndices] = {}
+        self._parse_file()
+
+    def _parse_file(self):
+        # Read offset and length for each data block
+        block_info: BinaryBlockInfo = SimulariumBinaryReader._parse_binary_header(
+            self.file_data.byte_view
+        )
+        for block_index in range(block_info.n_blocks):
+            block_type_id = SimulariumBinaryReader._binary_block_type(
+                block_index, block_info, self.file_data.int_view
+            )
+            self.block_indices[block_type_id] = DataIndices(
+                offset=block_info.block_offsets[block_index],
+                length=block_info.block_lengths[block_index],
+            )
+
+        # Extract each frame's metadata
+        spatial_indices: DataIndices = self.block_indices[
+            BINARY_BLOCK_TYPE.SPATIAL_DATA_BINARY.value
+        ]
+        spatial_block_offset = (
+            int(spatial_indices.offset / BINARY_SETTINGS.BYTES_PER_VALUE)
+            + BINARY_SETTINGS.BLOCK_HEADER_N_VALUES
+        )
+        n_frames = self.file_data.int_view[spatial_block_offset + 1]
+        current_frame_offset = spatial_block_offset + 2 + 2 * n_frames
+        for i in range(n_frames):
+            offset = (
+                self.file_data.int_view[spatial_block_offset + 2 + (2 * i)]
+                + spatial_indices.offset
+            )
+            length = self.file_data.int_view[spatial_block_offset + 3 + 2 * i]
```

Comment:
some of these "magic number" offsets of 2 or 3 could use either comments or symbolic constants.